float32 distance_forward